content/blog/2016/2016-08-10-rails-cd-with-pipeline.adoc

```diff
@@ -0,0 +1,358 @@
+        rvm 'bundle install' // <2>
+        rvm 'bundle exec rake'
+    } // <3>
+}
+----
+<1> Run the named container. The `inside` method can take optional additional flags for the `docker run` command.
+<2> Execute our shell commands using our tiny `sh` step wrapper
+link:https://github.com/rtyler/cfp-app/blob/aff11d97b460ca4b630b1fa4f669c226aeadd8ae/Jenkinsfile#L86-L91[`rvm`]. This ensures that the shell code is executed in the correct RVM environment.
+<3> When the closure completes, the container will be destroyed.
+
+
+
+Unfortunately, with this application, the `bundle exec rake` command will fail
+if PostgreSQL isn't available when the process starts. This is where the
+_second_ important feature of the CloudBees Docker Pipeline plugin comes
+into effect: the ability to run a container in the "background."
+
+[source, groovy]
+----
+node('docker') {
+    // .. 'stage' steps removed
+    /* Pull the latest `postgres` container and run it in the background */
+    docker.image('postgres').withRun { container -> // <1>
+        echo "PostgreSQL running in container ${container.id}" // <2>
+    } // <3>
+}
+----
+<1> Run the container, effectively `docker run postgres`
+<2> Any number of steps can go inside the closure
+<3> When the closure completes, the container will be destroyed.
+
+
+==== Running the tests
+
+Combining these two snippets of Jenkins Pipeline is, in my opinion, where the
+power of the link:https://en.wikipedia.org/wiki/Domain-specific_language[DSL]
+shines:
+
+[source,groovy]
+----
+node('docker') {
+    docker.image('postgres').withRun { container ->
+        docker.image('rtyler/rvm:2.3.0').inside("--link=${container.id}:postgres") { // <1>
+            stage 'Install Gems'
+            rvm "bundle install"
+
+            stage 'Invoke Rake'
+            withEnv(['DATABASE_URL=postgres://postgres@postgres:5432/']) { // <2>
+                rvm "bundle exec rake"
+            }
+            junit 'spec/reports/*.xml' // <3>
+        }
+    }
+}
+----
+<1> By passing the `--link` argument, the Docker daemon will allow the RVM container to talk to the PostgreSQL container under the host name 'postgres'.
+<2> Use the `withEnv` step to set environment variables for everything that is in the closure. In this case, the cfp-app DB scaffolding will look for the `DATABASE_URL` variable to override the DB host/user/dbname defaults.
+<3> Archive the test reports generated by link:https://github.com/ci-reporter/ci_reporter[ci_reporter] so that Jenkins can display test reports and trend analysis.
+
+
+image::/images/post-images/ruby-pipeline-2016/cfpapp-tests.png[role=center]
+
+With this done, the basics are in place to consistently run the tests for
+cfp-app in fresh Docker containers for each execution of the pipelinel.
+
+
+
+=== Security scanning
+
+Using link:http://brakemanscanner.org[Brakeman], the security scanner for Ruby
+on Rails, is almost trivially easy inside of Jenkins Pipeline. All courtesy of the
```

Comment:
You could combine this into one sentence, I think, and it reads a bit better. "...inside of Jenkins Pipeline, thanks to the Brakeman plugin, which implements..."